packages/itwinui-react/src/core/Table/Table.tsx

```diff
@@ -1128,3 +1137,34 @@ export const Table = <
 };
 
 export default Table;
+
+// ----------------------------------------------------------------------------
+
+/**
+ * Wrapper around `<template>` element that attaches shadow root to its parent
+ * and moves its children into the shadow root.
+ */
+const ShadowTemplate = ({ children }: { children: React.ReactNode }) => {
```

Comment:
i noticed two things:
1. empty `<template>` tag is left behind in the DOM
2. state updates from inside the template are not correctly reflected.

i was able to fix these using `template.remove()` and `createPortal()`

try these two snippets in the playground to see the difference:


<details><summary>Before</summary>

```tsx
import * as React from 'react';
import * as ReactDOM from 'react-dom';

export default function App() {
  const [count, setCount] = React.useState(0);

  return (
    <>
      <div>
        <ShadowTemplate>
          <button onClick={() => setCount(count + 1)}>{count}</button>
        </ShadowTemplate>
      </div>
    </>
  );
}

const ShadowTemplate = ({ children }: { children: React.ReactNode }) => {
  const [root, setRoot] = React.useState<HTMLDivElement | null>(null);

  const attachShadowRef = React.useCallback(
    (template: HTMLTemplateElement | null) => {
      const parent = template?.parentElement;
      if (!template || !root || parent?.shadowRoot) {
        return;
      }

      const shadowRoot = parent?.attachShadow({ mode: 'open' });
      shadowRoot?.appendChild(root);
    },
    [root],
  );

  return (
    <template ref={attachShadowRef}>
      <div ref={setRoot} style={{ display: 'contents' }}>
        {children}
      </div>
    </template>
  );
};

```

</details> 

<details><summary>After</summary>

```tsx
import * as React from 'react';
import * as ReactDOM from 'react-dom';

export default function App() {
  const [count, setCount] = React.useState(0);

  return (
    <>
      <div>
        <ShadowTemplate>
          <button onClick={() => setCount(count + 1)}>{count}</button>
        </ShadowTemplate>
      </div>
    </>
  );
}

const ShadowTemplate = ({ children }: { children: React.ReactNode }) => {
  const [root, setRoot] = React.useState<HTMLDivElement | null>(null);

  const attachShadowRef = React.useCallback(
    (template: HTMLTemplateElement | null) => {
      const parent = template?.parentElement;
      if (!template || !root || parent?.shadowRoot) {
        return;
      }

      const shadowRoot = parent?.attachShadow({ mode: 'open' });
      shadowRoot?.appendChild(root);

      template.remove();
    },
    [root],
  );

  return (
    <template ref={attachShadowRef}>
      <div ref={setRoot} style={{ display: 'contents' }}>
        {root && ReactDOM.createPortal(children, root)}
      </div>
    </template>
  );
};

```

</details>

Reply:
Ohh, I now understand why portal was needed, thanks! Added `template.remove()` & `createPortal` (https://github.com/iTwin/iTwinUI/pull/1725/commits/527ba95058f9228641a3208f636f41d57635b2f4)

Reply:
Ohh, sounds good, removed (https://github.com/iTwin/iTwinUI/pull/1725/commits/d722e44498297a227345ff2aac49caf76922cd67)